crates/adapters/src/format/csv.rs

```diff
@@ -80,129 +77,85 @@ impl CsvParser {
-            match result {
-                ReadRecordResult::End => break,
-                // `InputEmpty` status can be returned when there is no newline character in
-                // the end of the last input record, which isn't really an error.  So we treat
-                // it as success and leave it to the actual record parser to deal with possible
-                // invalid CSV (our job here is simply to establish record boundaries).
-                ReadRecordResult::Record | ReadRecordResult::InputEmpty => {
-                    /*println!(
-                        "result: {result:?}, record: '{}', bytes: {:?}",
-                        std::str::from_utf8(&record_buffer[0..total_bytes_read])
-                            .unwrap_or("invalid utf-8"),
-                        &record_buffer[0..total_bytes_read],
-                    );*/
-                    match self
-                        .input_stream
-                        .insert(&record_buffer[0..total_bytes_read])
-                    {
-                        Err(e) => {
-                            errors.push(ParseError::text_event_error(
-                                "failed to deserialize CSV record",
-                                e,
-                                self.last_event_number + 1,
-                                Some(
-                                    &std::str::from_utf8(&record_buffer[0..total_bytes_read])
-                                        .map(|s| s.to_string())
-                                        .unwrap_or_else(|_| {
-                                            format!("{:?}", &record_buffer[0..total_bytes_read])
-                                        })
-                                        .to_string(),
-                                ),
-                                None,
-                            ));
-                        }
-                        Ok(()) => {
-                            num_records += 1;
-                        }
-                    }
-                    // Lines ending in "\r\n" get broken up after `\r` by the parser.
-                    // Consume the remaining `\n`; otherwise it gets prepended to the
-                    // next buffer.
-                    if buffer.len() > bytes_read && buffer[bytes_read] == b'\n' {
-                        bytes_read += 1;
-                    }
-                    record_buffer = &buffer[bytes_read..];
-                    self.last_event_number += 1;
-                    total_bytes_read = 0;
-                    if result == ReadRecordResult::InputEmpty {
-                        break;
-                    }
+    /// Tries to split `buffer` into a CSV record followed by any remaining text.
+    ///
+    /// A CSV record is usually one line that ends in a new-line, but it can
+    /// span multiple lines if new-lines are enclosed in double quotes.
+    fn split_record<'a>(&mut self, buffer: &'a [u8]) -> Option<(&'a [u8], &'a [u8])> {
```

Comment:
Normally double quotes get escaped in CSV by doubling them, so that `"a""b""c"` unescapes to `a"b"c`. So, any even number of double quotes followed by a newline ends a record, but any odd number followed by a newline is a continuation of the field.

If we allow the user to configure the CSV format used, we'll need to adjust this to match the configuration.

Reply:
Nice! Maybe add a comment explaining this?

Reply:
I added a comment.